### PR TITLE
Add Conformalized Quantile Regression (ConformalRegCQR)

### LINKED ISSUE #325: :sparkles:(conformal): add Conformalized Quantile Regression (CQR)
### Motivation

The library already has a conformal module, but it's classification-only. 

### Description

CQR (Romano et al., 2019) is the canonical post-hoc calibration of quantile predictions to give finite-sample coverage guarantees.

### Paper reference (if applicable)

Romano, Y., Patterson, E., & Candes, E. (2019). Conformalized quantile regression. Advances in neural information processing systems, 32.

### Alternatives considered

_No response_

### Contribution

- [x] I would be willing to implement this feature and open a pull request.

### PR DESCRIPTION
**Pull request by kyb8801**
## Description

Adds Conformalized Quantile Regression (CQR; Romano et al., 2019) as a regression conformal post-processor, `ConformalRegCQR`, complementing the existing classification conformal predictors.

The existing `Conformal` base is classification-specific (softmax scores, prediction sets, temperature scaling), so `ConformalRegCQR` subclasses `PostProcessing` directly and lives alongside the classification predictors in `post_processing/conformal/`. The wrapped model outputs the lower and upper quantiles as a `(batch, 2)` tensor. `fit` stores the finite-sample `(1 - alpha)`-quantile `q_hat` of the two-sided conformity score `E = max(q_lo - y, y - q_hi)`, and `forward` returns the calibrated interval `[q_lo - q_hat, q_hi + q_hat]`.

Reference: Romano, Y., Patterson, E., & Candès, E. (2019). Conformalized Quantile Regression. NeurIPS 32.

Interval evaluation metrics (coverage/width, Winkler, ...) tracked in #327 can follow in a second PR.

Fixes #325

## Checklist

- [x] Branch name is not `main` or `dev`
- [x] Tests pass locally (`uv run pytest tests`)
- [x] Code is formatted (`uv run ruff format && uv run ruff check --fix`)
- [x] Code coverage is not reduced too much
- [x] New functions/classes have type annotations and docstrings
- [x] If a new method is implemented, a reference to the paper is added to the References page
- [x] Licensed code from external sources is explicitly attributed

### CODE REVIEW

File: tests/post_processing/test_conformal_reg_cqr.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        calibration_set = list(zip(quantiles, y, strict=True))
+        dl = DataLoader(calibration_set, batch_size=256)
+
+        conformal = ConformalRegCQR(alpha=0.1, model=nn.Identity())
+        conformal.fit(dl)
+
+        interval = conformal(quantiles)
+        coverage = ((y >= interval[:, 0]) & (y <= interval[:, 1])).float().mean()
+        # CQR restores marginal coverage close to the 1 - alpha target.
+        assert coverage >= 0.87
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply tests/post_processing/test_conformal_reg_cqr.py:53]: Won't do. It's fine for unit tests.

File: src/torch_uncertainty/post_processing/conformal/conformal_reg_cqr.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        assert self.model is not None
+        self.model.eval()
+
+        lowers, uppers, targets = [], [], []
+        for inputs, target in dataloader:
+            quantiles = self.model(inputs.to(self.device))
+            lowers.append(quantiles[:, 0])
+            uppers.append(quantiles[:, 1])
+            targets.append(target.to(self.device).reshape(-1))
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply src/torch_uncertainty/post_processing/conformal/conformal_reg_cqr.py:80]: We should probably enforce it at the start, but anyway. I'll use the newly added `check_interval_shapes` from utils.

File: src/torch_uncertainty/post_processing/conformal/conformal_reg_cqr.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        self.q_hat = torch.quantile(scores, level, interpolation="higher")
+        self.trained = True
+
+    @torch.no_grad()
+    def forward(self, inputs: Tensor) -> Tensor:
+        """Return the calibrated ``(batch, 2)`` ``[lower, upper]`` interval."""
+        quantiles = self.model(inputs.to(self.device))
+        lower = quantiles[:, 0] - self.quantile
+        upper = quantiles[:, 1] + self.quantile
+        return torch.stack((lower, upper), dim=-1)
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply src/torch_uncertainty/post_processing/conformal/conformal_reg_cqr.py:100]: The second check is not useful as `.fit` will have been called beforehand.

**Review by o-laurent — APPROVED** [review]

**Comment by o-laurent** [comment]
Thank you, @kyb8801. I'll merge dev into main ASAP and might release 0.12.2 with it later.
